### PR TITLE
Added towncrier as a github workflow

### PR DESCRIPTION
**Pull request by kc611**
As titled,

This PR adds towncrier as a Github workflow to check release notes for every PR. The action is intended to fail if there's no new release note, or skipped if there's a `skip_release_notes` label.

### CODE REVIEW

File: docs/source/upcoming_changes/README.rst [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+* ``feature``: Signifying a new feature.
+
+* ``bugfix``: Signifying a bug fix.
+
+* ``doc``: Signifying a documentation improvement.
+
+* ``removal``: Signifying a deprecation or removal of public API.
+
+* ``misc``: A ticket has been closed, but it is not of interest to users.
```
Comment (kc611) [inline docs/source/upcoming_changes/README.rst:24]: These are the default categories of changes provided by `towncrier`. 
(https://towncrier.readthedocs.io/en/stable/tutorial.html#creating-news-fragments)

cc @numba/fpoc; We should be putting together a custom category list of changes that are to be reflected in the auto-generated release notes.

Comment (kc611) [reply docs/source/upcoming_changes/README.rst:24]: Made the following list of categories adapted from NumPy:

```
* ``highlight``: Adds a highlight bullet point to use as a possibly highlight
  of the release.
* ``np_support``: Addition of new NumPy functionality.
* ``deprecation``: Changes existing code to emit a DeprecationWarning.
* ``future``: Changes existing code to emit a FutureWarning.
* ``expired``: Removal of a deprecated part of the API.
* ``compatibility``: A change which requires users to change code and is not
  backwards compatible. (Not to be used for removal of deprecated features.)
* ``cuda``: Changes in the CUDA API.
* ``new_feature``: New user facing features like ``kwargs``.
* ``improvement``: General improvements and edge-case changes which are
  not new features or compatibility related.
* ``performance``: Performance changes that should not affect other behaviour.
* ``change``: Other changes
```

**Review by apmasell — APPROVED** [review]
Looking most at GitHub Actions

File: docs/source/upcoming_changes/README.rst (hunk 1 above)
Comment (stuartarchibald) [reply docs/source/upcoming_changes/README.rst]: @kc611 Suggest raising this at the next maintainer/triage meeting to make sure that this list acceptable to all maintainers, also think that the general form of the release notes needs a general discussion/acceptance from the same group.

File: docs/source/developer/contributing.rst [hunk 2]
```diff
@@ -320,6 +320,14 @@ hints, as well as between Numba types---such as ``Dict`` or ``Literal``---and ``
 To mitigate the risk of confusion we use a naming convention by which objects of the built-in ``typing`` module are
 imported with an ``pt`` prefix. For example, ``typing.Dict`` is imported as ``from typing import Dict as ptDict``.
 
+Release Notes
+'''''''''''''
+
+Pull Requests that add significant user-facing modifications, they may need to be mentioned in the release notes.
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:326]:
```suggestion
Pull Requests that add significant user-facing modifications may need to be mentioned in the release notes.
```

File: docs/source/developer/contributing.rst [hunk 3]
```diff
@@ -320,6 +320,14 @@ hints, as well as between Numba types---such as ``Dict`` or ``Literal``---and ``
...
+To add a the release notes, you need to create a short ``.rst`` file with a summary and place it in
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:327]:
```suggestion
To add a release note, a short ``.rst`` file needs creating containing a summary of the change and it needs to be placed in 
```

File: docs/source/developer/contributing.rst [hunk 4]
```diff
@@ -320,6 +320,14 @@ hints, as well as between Numba types---such as ``Dict`` or ``Literal``---and ``
 To mitigate the risk of confusion we use a naming convention by which objects of the built-in ``typing`` module are
 imported with an ``pt`` prefix. For example, ``typing.Dict`` is imported as ``from typing import Dict as ptDict``.
 
+Release Notes
+'''''''''''''
+
+Pull Requests that add significant user-facing modifications, they may need to be mentioned in the release notes.
+To add a the release notes, you need to create a short ``.rst`` file with a summary and place it in 
+``docs/upcoming_changes``. The file ``docs/upcoming_changes/README.rst`` details the format
+and filename conventions.
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:329]:
```suggestion
and file naming conventions.
```

File: docs/upcoming_changes/README.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+If you are unsure what pull request type to use, don't hesitate to ask in your
+PR.
+
+You can install ``towncrier`` and run ``towncrier build --draft``
+if you want to get a preview of how your change will look in the final release
+notes.
+
+.. note::
+    This README was adapted from the NumPy changelog readme under the terms of
+    the MIT licence.
```
Comment (stuartarchibald) [inline docs/upcoming_changes/README.rst:40]: I think that this should be a comment in a style compliant with https://github.com/numba/numba/blob/e7a1e8315d74b0f92ba8b26a40bbb1f28db94ba2/LICENSES.third-party#L454-L461
?

File: docs/upcoming_changes/README.rst [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+Make sure to use full sentences with correct case and punctuation, and please
+try to use Sphinx intersphinx using backticks. The fragment should have a
+header line and an underline using ``========`` followed by description of
+your user-facing changes as they should appear in the relase notes.
+
+Each file should be named like ``<PULL REQUEST>.<TYPE>.rst``, where
+``<PULL REQUEST>`` is a pull request number, and ``<TYPE>`` is one of:
+
+* ``highlight``: Adds a highlight bullet point to use as a possibly highlight
+  of the release.
```
Comment (stuartarchibald) [inline docs/upcoming_changes/README.rst:17]:
```suggestion
* ``highlight``: Adds a highlighted bullet point to use as a possible highlight
  of the release.
```

File: docs/upcoming_changes/README.rst [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+* ``np_support``: Addition of new NumPy functionality.
+* ``deprecation``: Changes existing code to emit a DeprecationWarning.
```
Comment (stuartarchibald) [inline docs/upcoming_changes/README.rst:19]:
```suggestion
* ``deprecation``: Changes to existing code that will now emit a DeprecationWarning/PendingDeprecationWarning.
```

**Review by stuartarchibald — DISMISSED** [review]
Thanks for working on this @kc611, this should help with automating the release process. I've not looked at the GHA workflows as @apmasell has done so (thanks @apmasell!). I've also not tested this manually. I have however reviewed the remaining content, most of the suggestions are minor things, but as noted in another comment, I suggest raising this PR for discussion at development meetings next week so as to ensure other interested parties are aware and have the opportunity to comment. Thanks again!

**Comment by sklam** [comment]
Comments that need addressing:

- [ ] [MIT license?](https://github.com/numba/numba/pull/8792#discussion_r1163159599)
- [ ] [Update docs](https://github.com/numba/numba/pull/8792#discussion_r1163155388)

**Review by sklam — APPROVED** [review]

**Pull request merged**